ext/objspace/objspace_dump.c

```diff
@@ -691,6 +758,8 @@ objspace_dump_all(VALUE os, VALUE output, VALUE full, VALUE since)
         if (dc.roots) dump_append(&dc, "]}\n");
     }
 
+    rb_shape_each_shape(shape_i, &dc);
```

Comment:
One other way to implement this which would avoid needing the cursor or the shape callback is something like a static `objspace_dump_shapes` method which iterates over all shapes linearly:

```
for (shape_id_t i = 0; i < GET_VM()->next_shape_id; i++) {  
  rb_shape_t * shape = rb_shape_get_shape_by_id(i); 
  ....
}
```

Reply:
I think I prefer the current way, because it has somewhat better encapsulation. If for some reason you want to change how shapes are stored or start collecting unused shapes, you can do it all in `shape.c`.